ApplicationController.java

```
@@ -0,0 +1,470 @@
+        }
+    }
+
+    private void createDigitalIdentityRecord(CallContext callContext)
+    {
+        prepareDigitalIdentityRecord(callContext);
+        callContext.getDigitalIdentityRecord().save();
+        logger.info("{} was created in the database", callContext.getTid());
+    }
+
+    // assign the new device public key
+    private void updateDigitalIdentityRecord(CallContext callContext)
+    {
+        prepareDigitalIdentityRecord(callContext);
+        callContext.getDigitalIdentityRecord().update();
+        logger.info("{} was updated in the database", callContext.getTid());
+    }
+
+    private void prepareDigitalIdentityRecord(CallContext callContext)
+    {
+        DigitalIdentityModel record =
+                new DigitalIdentityModel(this.digitalIdentityStore,
+                                            callContext.getRequest(),
+                                            constructDigitalIdentity(callContext.getRequest()));
+
+        callContext.setDigitalIdentityRecord(record);
+    }
+
+    private void verifyToolId(String toolID) throws CloudValidationException
+    {
+        if (!DigitalIdentity.idIsValid(toolID)) {
+            ArrayList<ValidationError> errors = new ArrayList<>();
+            errors.add(new ValidationError("deviceAlternateId", DigitalIdentity.validIdHint()));
+            throw new CloudValidationException(errors);
+        }
+        logger.info("{} is the valid tool identifier", toolID);
+    }
+
+    private void lockTID(CallContext callContext)
+    {
+        DigitalIdentityLockModel lock =
+                new DigitalIdentityLockModel(this.digitalIdentityStore, callContext.getRequest());
+        try {
+            if (lock.obtain() == DigitalIdentityLockModel.LockResult.LOCKED_BY_OTHER)
+                throw new CloudBasicException(CONFLICT_RESOURCE_EXISTS,
+                        callContext.getTid() + " is currently being executed in another request");
+        } catch (DbStoreException pe) {
```

Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)